src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -986,16 +1014,36 @@ DeferErrorIfUnsupportedUnionQuery(Query *subqueryTree,
 			rightArgRTI = ((RangeTblRef *) rightArg)->rtindex;
 			rightArgSubquery = (Node *) rt_fetch(rightArgRTI,
 												 subqueryTree->rtable)->subquery;
-			if (HasReferenceTable(rightArgSubquery))
+			if (HasRecurringTuples(rightArgSubquery, &recurType))
 			{
-				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-									 "cannot push down this subquery",
-									 "Reference tables are not supported with union"
-									 " operator", NULL);
+				break;
 			}
 		}
 	}
 
+	if (recurType == RECURRING_TUPLES_REFERENCE_TABLE)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "cannot push down this subquery",
+							 "Reference tables are not supported with union operator",
+							 NULL);
+	}
+	else if (recurType == RECURRING_TUPLES_FUNCTION)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "cannot push down this subquery",
+							 "Table functions are not supported with union operator",
+							 NULL);
+	}
+	else if (recurType == RECURRING_TUPLES_EMPTY_JOIN_TREE)
```

Comment:
I hit a minor issue, probably some of the existing functions require minor refactorings:

```SQL
SELECT user_id FROM users_table as u1 WHERE  value_2 > (SELECT 10) AND  
value_2 < (SELECT foo.user_id FROM (SELECT user_id FROM users_table WHERE u1.user_id = user_id) as foo RIGHT OUTER JOIN (SELECT 20) as bar ON true) ORDER BY 1 ASC LIMIT 5;
```
If you run the above query, we should error out since `SELECT 20` is in the outer part an outer join. However, we allow it to pushdown since `ShouldRecurseForReferenceTableJoinChecks` returns false for that outer join. The check `if (list_length(subroot->join_rel_list) > 0)` returns true for `SELECT 20` query. It looks like we should change the logic there such that we ensure `relids` of the `RELOPT_JOINREL` is not null.

(We should probably rename `ShouldRecurseForReferenceTableJoinChecks` to `ShouldRecurseForRecurringTuplesJoinChecks` or such)


(Note that the following works fine with immutable functions)
```SQL
 SELECT user_id FROM users_table as u1 WHERE  
value_2 > (SELECT 10) AND 
value_2 < (SELECT foo.user_id FROM (SELECT user_id FROM users_table WHERE u1.user_id = user_id) as foo RIGHT OUTER JOIN (SELECT i FROM generate_series(1,10) i) as bar ON true) ORDER BY 1 ASC LIMIT 5;
ERROR:  cannot pushdown the subquery
DETAIL:  There exist a table function in the outer part of the outer join
```

Reply:
Makes sense, this happened even in simple join cases. Added a fix and proper regression tests.